File: 2021/day19.py
import sys
import itertools

DEBUG = 'debug' in sys.argv

if 'test' in sys.argv:
    data = open('test19.txt').readlines()
else:
    data = open('day19.txt').readlines()


scanners = []
for ln in data:
    ln = ln.rstrip()
    if not ln:
        continue
    if ln.startswith( '---' ):
        scanners.append( [] )
    else:
        scanners[-1].append( eval(ln) )


def sub( pta, ptb ):
    return pta[0]-ptb[0], pta[1]-ptb[1], pta[2]-ptb[2]

def add( pta, ptb ):
    return pta[0]+ptb[0], pta[1]+ptb[1], pta[2]+ptb[2]


def rotatex( point ):
    x,y,z = point
    return x, z, -y

def rotatey( point ):
    x,y,z = point
    return z, y, -x

def rotatez( point ):
    x,y,z = point
    return -y, x, z

# This returns a vector of the 24 possible rotations of this point.
# Move the "facing" direction through all 6 points, then rotate around
# that axis four times.

def allrotate( point ):
    for direction in (
        rotatex, rotatex, rotatex, rotatez, # now pointing up / down
        rotatey, rotatey, rotatey, rotatex, # now pointing back / front
        rotatez, rotatez, rotatez, rotatey  # now pointing left / right
    ) * 2:
        yield point
        point = direction(point)


# This function answers the musical question, for each pair of points
# in the two scan lists, if they lined up, how many other points would
# line up?

def find_matches( scan1, scan2 ):
    set1 = set(scan1)
    for pt1 in scan1:
        for pt2 in scan2:
            delta = sub(pt2,pt1)
            if len(set1.intersection((sub(pt2x,delta) for pt2x in scan2))) >= 12:
                return delta

def process(scanners):
    # Make all rotations of all scanner readings.
    # This is a 25x24 array.

    rotations = [
        list(zip(*list((list(allrotate(pt)) for pt in scanner)))) for scanner in scanners
    ]

    # This is the set of known beacons.

    beacons = set(rotations[0][0])

    placed = { 0: (0, (0,0,0)) }
    while len(placed) < len(rotations):
        for i, scan1 in enumerate(rotations):
            if i in placed:
                continue
            escape = False

            # scan1 is our candidate scanner.

            for j, (rot2,pt2) in placed.items():
                scan2 = rotations[j][rot2]

                # scan2 is a scanner whose position and rotation is now known.

                for possrot in range(24): 

                    # Does this rotation of scan1 line up with scan2?

                    result = find_matches( scan2, scan1[possrot] )
                    if not result:
                        continue
                    if DEBUG:
                        print( "Matched", i, "rot", possrot, "vs", j )

                    # We have a winner.  Add this to the set of knowns.
                    # We shift scan1 to match the coords of scan2, then
                    # shift that to our master coords.

                    beacons |= {*(
                        sub(sub(pt,result),pt2) for pt in scan1[possrot] 
                    )}

                    placed[i] = ( possrot, add(result,pt2) )
                    escape = True
                    break

                if escape:
                    break

    return beacons, placed

def part1(info):
    return len(info[0])

def part2(info):
    placed = info[1].values()
    # Find the largest distance between any two scanners (not beacons).
    maxx = 0
    for pt0 in placed:
        for pt1 in placed:
            maxx = max( maxx, sum(abs(i) for i in sub(pt0[1],pt1[1])))
    return maxx

def visoalize(beacons, placed):
    import numpy as np
    import matplotlib.pyplot as plt
    from mpl_toolkits import mplot3d

    beacons = np.array(list(beacons))

    scanners = -np.array( [k[1] for k in placed.values()] )

    ax = plt.axes(projection='3d')
    ax.scatter3D( beacons[:,0], beacons[:,1], beacons[:,2], c='blue', marker='o' )
    ax.scatter3D( scanners[:,0], scanners[:,1], scanners[:,2], c='red', marker='^' )
    plt.show()

info = process(scanners)

if 'vis' in sys.argv:
    visoalize(*info)
    sys.exit(0)

print( "Part 1:", part1(info))
print( "Part 2:", part2(info))

File: 2021/day19.cpp
#include <iostream>
#include <fstream>
#include <sstream>
#include <string>
#include <vector>
#include <set>
#include <map>

using namespace std;

#include "utils.h"

typedef vector<Point> PointList;
typedef set<Point> PointSet;
typedef map<int,int> Rotations;
typedef map<int,Point> Offsets;
typedef vector<PointList> ScannerSet;

bool DEBUG = false;

struct Information {
    PointSet beacons;
    Rotations rotations;
    Offsets offsets;
};


void read( 
    istream && ifs,
    ScannerSet & scanners
)
{
    scanners.clear();
    for( string line; getline(ifs, line); )
    {
        if( line.size() < 2 )
            continue;
        if( line.substr(0,3) == "---"  )
            scanners.resize( scanners.size() + 1 );
        else
        {
            char c;
            int i, j, k;
            istringstream fin( line );
            fin >> i >> c >> j >> c >> k;
            scanners.back().emplace_back( i, j, k );
        }
    }
}




// This returns a vector of the 24 possible rotations of this point.
// Move the "facing" direction through all 6 points, then rotate around
// that axis four times.

void allrotate( PointList & out, const Point in )
{
    Point pt = in;
    out.clear();

    for( int i = 0; i < 2; i++ )
        for( const char * axis = "xxxzyyyxzzzy"; *axis; axis++ )
        {
            out.push_back( pt );
            pt.rotate(*axis);
        }
}


// This function answers the musical question, for each pair of points
// in the two scan lists, if they lined up, how many other points would
// line up?

bool find_matches( PointList scan1, PointList scan2, Point & delta )
{
    PointSet set1( scan1.begin(), scan1.end() );
    for( auto && pt1 : scan1 )
        for( auto && pt2 : scan2 )
        {
            int matches = 0;
            delta = pt2 - pt1;
            for( auto && pt2x : scan2 )
                if( set1.find( pt2x - delta ) != set1.end() )
                    matches += 1;
            if( matches >= 12 )
                return true;
        }
    return false;
}


bool process( ScannerSet scanners, /*out*/ Information & info )
{
    // Make all rotations of all scanner readings.
    // This is a 25x24 array.
    // rotations[i][j] is the set of points for scanner i in rotation j.

    vector< vector< PointList > > rotations;

    rotations.resize( scanners.size() );
    for( auto && r : rotations )
        r.resize( 24 );
    
    PointList rots;
    for( int i = 0; i < scanners.size(); i++ )
    {
        for( auto && pt : scanners[i] )
        {
            allrotate( rots, pt );
            for( int j = 0; j < rots.size(); j++ )
                rotations[i][j].push_back( rots[j] );
        }
    }

    PointSet & beacons = info.beacons;
    Rotations & known_r = info.rotations;
    Offsets & known_o = info.offsets;

    // This is the set of known beacons.

    beacons = PointSet( rotations[0][0].begin(), rotations[0][0].end() );
    known_r[0] = 0;
    known_o[0] = Point(0,0,0);

    while( known_r.size() < rotations.size() )
    {
        for( int i = 0; i < rotations.size(); i++ )
        {
            if( known_r.find(i) != known_r.end() )
                continue;

            vector<PointList> & scan1 = rotations[i];

            bool escape = false;

            // scan1 is the candidate scanner we want to place.

            for( auto && rot2 : known_r )
            {
                int j = rot2.first;
                auto & pt2 = known_o[j];
                auto & scan2 = rotations[j][rot2.second];

                // scan2 is a scanner whose position and rotation is now known.

                for( int possrot = 0; possrot < 24; possrot++ )
                {
                    // Does this rotation of scan1 line up with scan2?

                    Point potential;
                    if( !find_matches( scan2, scan1[possrot], potential ) )
                        continue;

                    if( DEBUG )
                    {
                        cout << "Matched " <<  i << "  rot " << possrot << " vs " << j << "\n";
                    }

                    // We have a winner.  Add this to the set of knowns.
                    // We shift scan1 to match the coords of scan2, then
                    // shift that to our master coords.

                    for( auto && pt : scan1[possrot] )
                    {
                        beacons.insert( (pt - potential) - pt2 );
                    }

                    known_r[i] = possrot;
                    known_o[i] = potential + pt2;
                    escape = true;
                    break;
                }

                if( escape )
                    break;
            }
        }
    }

    return true;
}

int part1( const Information & info )
{
    return info.beacons.size();
}

int part2( const Information & info )
{
    int maxx = 0;
    for( auto && pt1 : info.offsets )
        for( auto && pt2 : info.offsets )
        {
            int dist = pt1.second.mandist( pt2.second );
            if( dist > maxx )
                maxx = dist;
        }
    return maxx;
}

int main( int argc, char ** argv )
{
    const char * filename = "day19.txt";
    while( *++argv )
    {
        if( strcmp(*argv, "debug") == 0 )
            DEBUG = true;
        else if( strcmp(*argv, "test") == 0 )
            filename = "test19.txt";
    }

    vector<PointList> scanners;
    read( ifstream(filename), scanners );

    Information info;
    process(scanners, info);

    cout << "Part 1: " << part1(info) << "\n";
    cout << "Part 2: " << part2(info) << "\n";
    return 0;
}
